File: cf/571/e.cpp
#include <cassert>
#include <algorithm>
#include <cstring>
#include <cstdlib>
#include <cstdio>
#include <string>
#include <vector>
using namespace std;

#define all(a) a.begin(), a.end()
#define rep(i, a, b) for(int i=int(a); i<=int(b); ++i)
#define irep(i, a, b) for(int i=int(a); i>=int(b); --i)
#define repn(i, n) rep(i, 0, (n)-1)
#define irepn(i, n) irep(i, (n)-1, 0)
#define fillchar(a, x) memset(a, x, sizeof(a))
#define fi first
#define se second
#define pb push_back
#define mp make_pair
#ifdef DEBUG
    #define _debug(args...) { fprintf(stderr, args); fflush(stderr); }
#else
    #define _debug(args...) {}
#endif
typedef long long LL;
typedef pair<LL,LL> PLL;
typedef pair<int,int> PII;
typedef pair<double,double> PDD;
typedef vector<LL> VL;
typedef vector<int> VI;
typedef vector<PII> VPI;
typedef vector<string> VS;
template<class T> bool setmax(T &_a, T _b) { if(_b>_a) { _a=_b; return true; } return false; }
template<class T> bool setmin(T &_a, T _b) { if(_b<_a) { _a=_b; return true; } return false; }
template<class T> T gcd(T _a, T _b) { return _b==0?_a:gcd(_b,_a%_b); }
template<class T> T ext_gcd(T a, T b, T &x, T &y) {
    if(b==0) { x=1, y=0; return a; }
    LL nx, ny, d; d=ext_gcd(b, a%b, nx, ny);
    x=ny, y=nx-(a/b)*ny;
    return d;
}

const LL MOD=LL(1e9)+7;

bool parallel(LL x1, LL y1, LL x2, LL y2) { //{{{
    return x1*y2==x2*y1;
    /*
    if(x1==0 && y1==0) return true;
    if(x2==0 && y2==0) return true;
    LL g1=gcd(x1, y1), g2=gcd(x2, y2);
    x1/=g1, y1/=g1, x2/=g2, y2/=g2;
    if(x1+y1<0) x1=-x1, y1=-y1;
    if(x2+y2<0) x2=-x2, y2=-y2;
    return x1==x2 && y1==y2;
    */
} //}}}

struct Solution;
Solution f(LL a1, LL a2, LL b1, LL b2);

struct Solution { //{{{
    enum {
        MULTI, SINGLE, NONE, ALL
    } type;
    LL i, j, di, dj;

    Solution() 
        : type(NONE) {}
    Solution(bool b) 
        : type(b?ALL:NONE) {}
    Solution(LL i, LL j) 
        : type(SINGLE), i(i), j(j) {}
    Solution(LL i, LL j, LL di, LL dj) 
        : type(MULTI), i(i), j(j), di(di), dj(dj) { 
            assert(di>0 || dj>0); }

    Solution proj() const {
        if(type==MULTI) {
            if(di>0) return Solution(i, 0, di, 0);
            else return Solution(i, 0);
        }
        if(type==SINGLE) return Solution(i, 0);
        if(type==ALL) return Solution(0, 0, 1, 0);
        return Solution();
    }
    bool contains(LL x, LL y) const {
        if(type==ALL) return true;
        if(type==NONE) return false;
        if(type==SINGLE) return x==i && y==j;

        if(i>x || j>y) return false;
        LL dx=x-i, dy=y-j;
        if(!parallel(di, dj, dx, dy)) return false;
        LL u=dx+dy, v=di+dj;
        return u%v==0 && u/v>=0;
    }

    friend Solution intersect(Solution s1, Solution s2) {
        if(s1.type==NONE || s2.type==NONE) return Solution();
        if(s1.type==ALL) return s2;
        if(s2.type==ALL) return s1;
        if(s1.type==SINGLE) {
            if(s2.contains(s1.i, s1.j)) return s1;
            return Solution();
        }
        if(s2.type==SINGLE) {
            if(s1.contains(s2.i, s2.j)) return s2;
            return Solution();
        }
        // now MULTI
        if(parallel(s1.di, s1.dj, s2.di, s2.dj)) {
            LL dx=s1.i-s2.i, dy=s1.j-s2.j;
            if(!parallel(s1.di, s1.dj, dx, dy)) return Solution();
            Solution tmp;
            if(s1.di>0) {
                tmp=f(s1.i, s2.i, s1.di, s2.di);
            } else {
                tmp=f(s1.j, s2.j, s1.dj, s2.dj);
            }
            return Solution(s1.i+s1.di*tmp.i, s1.j+s1.dj*tmp.i, s1.di*tmp.di, s1.dj*tmp.di);
        } else {
            // must has one intersection
            auto make_line=[](LL x1, LL y1, LL x2, LL y2, LL &a, LL &b, LL &c) {
                a=y2-y1, b=x1-x2, c=-(a*x1+b*y1);
            };
            LL a1, b1, c1; make_line(s1.i, s1.j, s1.i+s1.di, s1.j+s1.dj, a1, b1, c1);
            LL a2, b2, c2; make_line(s2.i, s2.j, s2.i+s2.di, s2.j+s2.dj, a2, b2, c2);
            LL t=a1*b2-a2*b1, x=b1*c2-c1*b2, y=c1*a2-a1*c2;
            _debug("    intersect: multi, not parallel - t=%lld,x=%lld,y=%lld\n",t,x,y);
            assert(t!=0);
            if(x%t!=0 || y%t!=0) return Solution();
            x/=t, y/=t;
            _debug("    intersect: x=%lld y=%lld\n",x,y);
            if(s1.contains(x, y) && s2.contains(x, y))
                return Solution(x, y);
            return Solution();
        }
    }
}; //}}}

// (a1,b1), (a2,b2) -> (i,j) s.t. a1+b1*i = a2+b2*j
Solution f(LL a1, LL a2, LL b1, LL b2) { //{{{
    _debug("    f (%lld %lld) (%lld %lld)\n", a1,b1,a2,b2);
    if(b1==0 && b2==0) return Solution(a1==a2);
    if(b1==0) {
        if((a1-a2)%b2!=0 || (a1-a2)/b2<0) 
            return Solution();
        return Solution(0, (a1-a2)/b2, 1, 0);
    }
    if(b2==0) {
        if((a2-a1)%b1!=0 || (a2-a1)/b1<0)
            return Solution();
        return Solution((a2-a1)/b1, 0, 0, 1);
    }

    // now a1 a2 >= 0, b1 b2 >= 1
    LL i0, j0, g=ext_gcd(b1, b2, i0, j0); j0=-j0;
    if((a2-a1)%g!=0)
        return Solution();
    LL d=(a2-a1)/g; i0*=d, j0*=d;
    // now a1+b1*i0 = a2+b2*j0
    LL di=b2/g, dj=b1/g; // (i,j)+k(di,dj) is still solution

    LL t;
    t=max((-i0-1)/di, (-j0-1)/dj)+1;
    if(t>0) i0+=t*di, j0+=t*dj;
    t=min(i0/di, j0/dj); 
    if(t>0) i0-=t*di, j0-=t*dj;

    return Solution(i0, j0, di, dj);
} //}}}

// (a1,b1), (a2,b2) -> (i,j) s.t. a1*b1^i = a2*b2^j
Solution go(LL a1, LL a2, LL b1, LL b2) { //{{{
    _debug("go (%lld %lld) (%lld %lld)\n", a1,b1,a2,b2);
    auto factorize=[](LL n) {
        VL vec;
        for(LL x=2; x*x<=n; ++x)
            if(n%x==0) {
                vec.pb(x);
                while(n%x==0) n/=x;
            }
        if(n>1) vec.pb(n);
        return vec;
    };
    auto to_vec=[](const VL &ps, LL n) {
        VL vec;
        for(LL p: ps) {
            LL cur=0; 
            while(n%p==0) ++cur, n/=p;
            vec.pb(cur);
        }
        return vec;
    };

    VL primes, tmp;
    tmp=factorize(a1); primes.insert(primes.end(), all(tmp));
    tmp=factorize(a2); primes.insert(primes.end(), all(tmp));
    tmp=factorize(b1); primes.insert(primes.end(), all(tmp));
    tmp=factorize(b2); primes.insert(primes.end(), all(tmp));
    sort(all(primes)); 
    primes.erase(unique(all(primes)), primes.end());

    VL x1=to_vec(primes, a1);
    VL x2=to_vec(primes, a2);
    VL y1=to_vec(primes, b1);
    VL y2=to_vec(primes, b2);
    Solution sol(true);
    repn(i, primes.size()) {
        _debug("  f_begin prime=%lld\n",primes[i]);
        Solution tmp=f(x1[i], x2[i], y1[i], y2[i]);
        _debug("  f_end tmp=(%d,%lld,%lld,%lld,%lld)\n",tmp.type,tmp.i,tmp.j,tmp.di,tmp.dj);
        sol=intersect(sol, tmp);
        _debug("  intersect_end sol=(%d,%lld,%lld,%lld,%lld)\n",sol.type,sol.i,sol.j,sol.di,sol.dj);
    }
    _debug("go_end sol=(%d,%lld,%lld,%lld,%lld)\n",sol.type,sol.i,sol.j,sol.di,sol.dj);
    return sol;
} //}}}

LL solve(VPI as) { //{{{
    if(as.size()==1) return as[0].fi;

    _debug("solve #1\n");

    Solution sol(true); // all
    for(PII a: as) {
        Solution tmp=go(as[0].fi, a.fi, as[0].se, a.se).proj();
        sol=intersect(sol, tmp);
    }

    _debug("solve #2\n");

    LL k=0;
    if(sol.type==Solution::NONE)
        return -1;
    if(sol.type==Solution::ALL) k=0;
        else k=sol.i;

    auto pow=[](LL a, LL b) {
        LL c=1;
        for(; b>0; b>>=1, a=a*a%MOD)
            if(b&1) c=c*a%MOD;
        return c;
    };
    LL ans=as[0].fi*pow(as[0].se, k)%MOD;
    return ans;
} //}}}

int main() {
    int n; scanf("%d", &n);
    VPI as(n); for(PII &p: as) scanf("%d%d", &p.fi,&p.se);
    LL ans=solve(as);
    printf("%d\n", (int)ans);
    return 0;
}

File: cf/571/e.hs
import Data.Functor
import Data.List (nub, zip4)
import Data.Bits

type Pair = (Integer, Integer)
modNum = 1000000007 :: Integer
modIt :: Integer -> Integer
modIt a = mod a modNum
modMul :: Integer -> Integer -> Integer
modMul a b = modIt $ a * b
modPow :: Integer -> Integer -> Integer
modPow a b = if b == 0 then 1 else modIt $
    (if b .&. 1 == 1 then a else 1) * modPow (modMul a a) (div b 2)
divides :: Integer -> Integer -> Bool
divides a b = (mod b a) == 0

extGCD :: Integer -> Integer -> (Integer, Integer, Integer)
extGCD a 0 = (a, 1, 0)
extGCD a b = (d, ny, nx - (a `div` b) * ny)
    where (d, nx, ny) = extGCD b (a `mod` b)

main :: IO ()
main = getInts >>= putStrLn . show . solve . parsePairs . tail where
    getInts = map read <$> words <$> getContents
    parsePairs [] = []
    parsePairs (x:y:rs) = (x, y) : parsePairs rs
--main = putStrLn . show $ modPow (123::Integer) (321::Integer)
--main = putStrLn . show $ factorize (modNum::Integer)

---------------------------------------------------------------------
data Solution = Multi Pair Pair | Single Pair | None | All
getSrc (Multi p _) = p
getDir (Multi _ p) = p

projToX :: Solution -> Solution
projToX s = case s of Multi (i, _) (di, _) -> Multi (i, 0) (di, 0)
                      Single (i, _) -> Single (i, 0)
                      None -> None
                      All -> Multi (0, 0) (1, 0)

simplify i0 j0 di dj = Multi (i0'', j0'') (di, dj) -- TODO
    where
        t1 = max 0 $ max (div (-i0-1) di) (div (-j0-1) dj) + 1
        i0' = i0 + t1 * di
        j0' = j0 + t1 * dj
        t2 = max 0 $ min (div i0 di) (div j0 dj)
        i0'' = i0' - t2 * di
        j0'' = j0' - t2 * dj

contains :: Pair -> Solution -> Bool
contains _ _ = False -- TODO

parallel :: Pair -> Pair -> Bool
parallel (x1, y1) (x2, y2) = x1 * y2 == x2 * y1

intersect :: Solution -> Solution -> Solution
intersect None _ = None
intersect _ None = None
intersect a All = a
intersect All a = a
intersect (Single p) a = if contains p a then Single p else None
intersect a (Single p) = if contains p a then Single p else None
intersect (Multi s1 d1) (Multi s2 d2) = if parallel d1 d2 then func1 else func2
    where
        func1 = if not (parallel (dx dy) d1) then None else val
            where (dx, dy) = (fst s1 - fst s2, snd s1 - snd s2)
                  tmp = if fst d1 > 0 then f (fst s1, fst d1) (fst s2, fst d2)
                                      else f (snd s1, snd d1) (snd s2, snd d2)
                  val = case tmp of Multi
        func2 = 

---------------------------------------------------------------------
solve :: [Pair] -> Integer
solve (p:[]) = modIt $ fst p
solve (p:rs) = case sol of Multi (i, _) (_, _) -> answer i
                           Single (i, _) -> answer i
                           All -> answer 0
                           None -> -1
    where sol = foldl intersect All . map projToX $ map (go p) rs
          answer i = modIt $ fst p * modPow (snd p) i

---------------------------------------------------------------------
-- (a1,b1), (a2,b2) -> (i,j) s.t. a1*b1^i = a2*b2^j
go :: Pair -> Pair -> Solution
go (a1, b1) (a2, b2) = foldl intersect All . map f' $ zip4 a1' b1' a2' b2'
    where primes = nub (factorize a1 ++ factorize a2 ++ factorize b1 ++ factorize b2)
          [a1', b1', a2', b2'] = map (toVec primes) [a1, b1, a2, b2]
          f' (x1, y1, x2, y2) = f (x1, y1) (x2, y2)

factorize :: Integer -> [Integer]
factorize 1 = []
factorize n = let xs = filter (`divides` n) [2 .. floor . sqrt $ fromIntegral n] in
    if xs == [] then [n] else let x = head xs in x:factorize (div n x)

toVec [] n = []
toVec (p:rs) n = getNum p n : toVec rs n where
    getNum p n = if p `divides` n then 1 + getNum p (div n p) else 0

---------------------------------------------------------------------
-- (a1,b1), (a2,b2) -> (i,j) s.t. a1+b1*i = a2+b2*j
f :: Pair -> Pair -> Solution
f (a1, 0) (a2, 0) = if a1 == a2 then All else None
f (a1, 0) (a2, b2) = let (x, y) = (a1-a2, b2) in 
    if mod x y /= 0 || div x y < 0 then None else Multi (0, div x y) (1, 0)
f (a1, b1) (a2, 0) = let (x, y) = (a2-a1, b1) in 
    if mod x y /= 0 || div x y < 0 then None else Multi (div x y, 0) (0, 1)
f (a1, b1) (a2, b2) = if (a2 - a1) `mod` g == 0 then val else None
    where 
        (g, i0', j0') = extGCD b1 (-b2)
        d = (a2 - a1) `div` g
        (i0, j0) = (i0' * d, j0' * d)
        (di, dj) = (b2 `div` g, b1 `div` g)
        val = simplify i0 j0 di dj

---------------------------------------------------------------------

